File: backend/utils/logger.ts
/**
 * Production-ready logging utility
 * Provides structured logging with different levels and contexts
 */

export enum LogLevel {
    DEBUG = 'DEBUG',
    INFO = 'INFO',
    WARN = 'WARN',
    ERROR = 'ERROR'
}

export interface LogContext {
    component?: string;
    operation?: string;
    server?: string;
    tool?: string;
    userId?: string;
    requestId?: string;
    duration?: number;
    [key: string]: any;
}

class Logger {
    private logLevel: LogLevel;
    private enableConsole: boolean;
    private enableFile: boolean;
    private logs: any[] = [];

    constructor() {
        this.logLevel = this.parseLogLevel(process.env.LOG_LEVEL || 'INFO');
        this.enableConsole = process.env.LOG_CONSOLE !== 'false';
        this.enableFile = process.env.LOG_FILE === 'true';
    }

    private parseLogLevel(level: string): LogLevel {
        const upperLevel = level.toUpperCase();
        return LogLevel[upperLevel as keyof typeof LogLevel] || LogLevel.INFO;
    }

    private shouldLog(level: LogLevel): boolean {
        const levels = [LogLevel.DEBUG, LogLevel.INFO, LogLevel.WARN, LogLevel.ERROR];
        const currentLevelIndex = levels.indexOf(this.logLevel);
        const messageLevelIndex = levels.indexOf(level);
        return messageLevelIndex >= currentLevelIndex;
    }

    private formatLog(level: LogLevel, message: string, context?: LogContext): string {
        const timestamp = new Date().toISOString();
        const contextStr = context ? ` | ${JSON.stringify(context)}` : '';
        return `[${timestamp}] [${level}] ${message}${contextStr}`;
    }

    private writeLog(level: LogLevel, message: string, context?: LogContext, error?: Error) {
        if (!this.shouldLog(level)) return;

        const logEntry = {
            timestamp: new Date().toISOString(),
            level,
            message,
            context,
            error: error ? {
                name: error.name,
                message: error.message,
                stack: error.stack
            } : undefined
        };

        // Store in memory for retrieval
        this.logs.push(logEntry);
        if (this.logs.length > 1000) {
            this.logs.shift(); // Keep only last 1000 logs
        }

        // Console output with colors
        if (this.enableConsole) {
            const formatted = this.formatLog(level, message, context);
            switch (level) {
                case LogLevel.DEBUG:
                    console.debug('\x1b[36m%s\x1b[0m', formatted); // Cyan
                    break;
                case LogLevel.INFO:
                    console.info('\x1b[32m%s\x1b[0m', formatted); // Green
                    break;
                case LogLevel.WARN:
                    console.warn('\x1b[33m%s\x1b[0m', formatted); // Yellow
                    break;
                case LogLevel.ERROR:
                    console.error('\x1b[31m%s\x1b[0m', formatted); // Red
                    if (error?.stack) {
                        console.error('\x1b[31m%s\x1b[0m', error.stack);
                    }
                    break;
            }
        }

        // File output (if enabled)
        if (this.enableFile) {
            // In production, you would write to a file or send to a logging service
            // For now, we'll just keep in memory
        }
    }

    debug(message: string, context?: LogContext) {
        this.writeLog(LogLevel.DEBUG, message, context);
    }

    info(message: string, context?: LogContext) {
        this.writeLog(LogLevel.INFO, message, context);
    }

    warn(message: string, context?: LogContext) {
        this.writeLog(LogLevel.WARN, message, context);
    }

    error(message: string, context?: LogContext, error?: Error) {
        this.writeLog(LogLevel.ERROR, message, context, error);
    }

    // Get recent logs for debugging
    getRecentLogs(count: number = 100): any[] {
        return this.logs.slice(-count);
    }

    // Clear logs
    clearLogs() {
        this.logs = [];
    }
}

// Singleton instance
export const logger = new Logger();

// Performance monitoring utility
export class PerformanceMonitor {
    private metrics: Map<string, number[]> = new Map();

    startTimer(operation: string): () => number {
        const start = Date.now();
        return () => {
            const duration = Date.now() - start;
            this.recordMetric(operation, duration);
            return duration;
        };
    }

    recordMetric(operation: string, duration: number) {
        if (!this.metrics.has(operation)) {
            this.metrics.set(operation, []);
        }
        const metrics = this.metrics.get(operation)!;
        metrics.push(duration);

        // Keep only last 100 measurements
        if (metrics.length > 100) {
            metrics.shift();
        }

        logger.debug(`Performance metric recorded`, {
            operation,
            duration,
            unit: 'ms'
        });
    }

    getStats(operation: string): { avg: number; min: number; max: number; count: number } | null {
        const metrics = this.metrics.get(operation);
        if (!metrics || metrics.length === 0) return null;

        return {
            avg: metrics.reduce((a, b) => a + b, 0) / metrics.length,
            min: Math.min(...metrics),
            max: Math.max(...metrics),
            count: metrics.length
        };
    }

    getAllStats(): Record<string, any> {
        const stats: Record<string, any> = {};
        for (const [operation, _] of this.metrics) {
            stats[operation] = this.getStats(operation);
        }
        return stats;
    }
}

export const perfMonitor = new PerformanceMonitor();

File: backend/utils/logger.py
import logging
import json
import os
import time
from datetime import datetime
from enum import Enum
from typing import Any, Dict, List, Optional
from uuid import uuid4

class LogLevel(str, Enum):
    DEBUG = "DEBUG"
    INFO = "INFO"
    WARN = "WARN"
    ERROR = "ERROR"

class Logger:
    def __init__(self):
        log_level_str = os.getenv("LOG_LEVEL", "INFO").upper()
        self.log_level = getattr(logging, log_level_str, logging.INFO)
        self.enable_console = os.getenv("LOG_CONSOLE", "true").lower() == "true"
        
        self.logs = []
        self.max_logs = 1000
        
        # Configure standard logging
        logging.basicConfig(
            level=self.log_level,
            format='%(message)s'
        )
        self.logger = logging.getLogger("mcp_agent")

    def _format_log(self, level: LogLevel, message: str, context: Optional[Dict[str, Any]] = None, error: Optional[Exception] = None) -> Dict[str, Any]:
        log_entry = {
            "timestamp": datetime.utcnow().isoformat() + "Z",
            "level": level,
            "message": message,
            "context": context or {},
        }
        
        if error:
            log_entry["error"] = {
                "name": type(error).__name__,
                "message": str(error),
                # stack trace could be added here if needed
            }
            
        return log_entry

    def _write_log(self, level: LogLevel, message: str, context: Optional[Dict[str, Any]] = None, error: Optional[Exception] = None):
        log_entry = self._format_log(level, message, context, error)
        
        # Store in memory
        self.logs.append(log_entry)
        if len(self.logs) > self.max_logs:
            self.logs.pop(0)
            
        # Console output
        if self.enable_console:
            color = ""
            if level == LogLevel.DEBUG: color = "\033[36m" # Cyan
            elif level == LogLevel.INFO: color = "\033[32m" # Green
            elif level == LogLevel.WARN: color = "\033[33m" # Yellow
            elif level == LogLevel.ERROR: color = "\033[31m" # Red
            
            reset = "\033[0m"
            context_str = f" | {json.dumps(context)}" if context else ""
            print(f"{color}[{log_entry['timestamp']}] [{level}] {message}{context_str}{reset}")
            if error:
                print(f"{color}{str(error)}{reset}")

    def debug(self, message: str, context: Optional[Dict[str, Any]] = None):
        self._write_log(LogLevel.DEBUG, message, context)

    def info(self, message: str, context: Optional[Dict[str, Any]] = None):
        self._write_log(LogLevel.INFO, message, context)

    def warn(self, message: str, context: Optional[Dict[str, Any]] = None):
        self._write_log(LogLevel.WARN, message, context)

    def error(self, message: str, context: Optional[Dict[str, Any]] = None, error: Optional[Exception] = None):
        self._write_log(LogLevel.ERROR, message, context, error)

    def get_recent_logs(self, count: int = 100) -> List[Dict[str, Any]]:
        return self.logs[-count:]

logger = Logger()

class PerformanceMonitor:
    def __init__(self):
        self.metrics = {}

    def start_timer(self, operation: str):
        start_time = time.time()
        def end_timer():
            duration = (time.time() - start_time) * 1000 # to ms
            self.record_metric(operation, duration)
            return duration
        return end_timer

    def record_metric(self, operation: str, duration: float):
        if operation not in self.metrics:
            self.metrics[operation] = []
        
        self.metrics[operation].append(duration)
        if len(self.metrics[operation]) > 100:
            self.metrics[operation].pop(0)
            
        logger.debug(f"Performance metric recorded", {"operation": operation, "duration": duration, "unit": "ms"})

    def get_stats(self, operation: str) -> Optional[Dict[str, Any]]:
        metric_list = self.metrics.get(operation)
        if not metric_list:
            return None
            
        return {
            "avg": sum(metric_list) / len(metric_list),
            "min": min(metric_list),
            "max": max(metric_list),
            "count": len(metric_list)
        }

    def get_all_stats(self) -> Dict[str, Any]:
        return {op: self.get_stats(op) for op in self.metrics}

perf_monitor = PerformanceMonitor()
